Adds support for in-app purchases

hannahtinkler opened the pull request:
Disables purchase management and auto-renew toggle for in-app purchases, adds a user notice regarding how to manage the purchase (via the app store) and adds 'In-App Purchase' as a visual payment method in the purchase meta.

Fixes #58496

#### Changes proposed in this Pull Request
* Remove purchase management options if the API returns `purchase.is_locked: true`
* Remove 'Upgrade' button from the plans page if `purchase.is_locked: true` for that plan purchase
* Show 'In-App Purchase' under payment method if the API returns `purchase.is_iap_purchase: true`
* Show 'This product is an in-app purchase, it can only be managed via the app store.' notice if API returns `purchase.is_locked: true` and `purchase.is_iap_purchase: true`

#### Testing instructions
* Apply diff D70663 for the public API changes
* Have a (non-ecommerce) subscription on your account that was paid for using in-app-purchases (payment method is 'iap_ios', or ping me and I can add one to your account).
* View the purchase in Calypso (e.g. `/me/purchases/{ site_url }/{ purchase_id }`)
* Assert that there are no options to manage the purchase (e.g. renew, refund, toggle auto-renew, cancel, remove etc)
* Assert that the payment method shows 'In-App Purchase'
* Assert that there is a user notice instructing the user to manage the purchase in the app store
* Assert that there is no 'Upgrade' button on the Plans page (`/plans/{ site_url }`)
* Have a subscription on your account that was **not** paid for using in-app-purchases (e.g. added via Store Admin).
* View the purchase in Calypso (e.g. `/me/purchases/{ site_url }/{ purchase_id }`)
* Assert that there are the usual options to manage the purchase (e.g. renew, refund, toggle auto-renew, cancel, remove etc)
* Assert that the payment method shows the expected value (e.g. 'None' for a Store Admin addition)
* Assert that there is not a user notice instructing the user to manage the purchase in the app store

hannahtinkler committed:
Adds support for in-app purchases

Disables purchase management and auto-renew toggle for in-app purchases, adds a user notice regarding how to manage the purchase (via the app store) and adds 'In-App Purchase' as a visual payment method in the purchase meta.

Relates to #58496

hannahtinkler committed:
Adds support for in-app purchases

Disables purchase management and auto-renew toggle for in-app purchases, adds a user notice regarding how to manage the purchase (via the app store) and adds 'In-App Purchase' as a visual payment method in the purchase meta.

Relates to #58496

hannahtinkler commented:
### Visual changes
Before:
![Screenshot 2021-11-25 at 12 32 30](https://user-images.githubusercontent.com/7140719/143472745-6794187b-f917-4240-b0a7-d7bae4e1b08a.png)

After:
![Screenshot 2021-11-25 at 12 34 03](https://user-images.githubusercontent.com/7140719/143472802-93613a8c-ac75-4d49-a446-529ef7c7bf9e.png)

This applies to IAP purchases only, non-IAP purchases (e.g. Stripe, Paypal etc) should show no visual difference.

a8ci18n commented on client/me/purchases/manage-purchase/payment-info-block.tsx:
```diff
@@ -69,6 +69,10 @@ export default function PaymentInfoBlock( { purchase }: { purchase: Purchase } )
 	}
 
+	if ( purchase.isInAppPurchase ) {
+		return <PaymentInfoBlockWrapper>{ translate( 'In-App Purchase' ) }</PaymentInfoBlockWrapper>;
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **2** times:
`translate( 'Apple In App Purchase' )` *ES Score: 8*

hannahtinkler replied:
This could be an Apple or Android in-app purchase 😞 So whilst we could make this app store-specific (in order to reuse the string), it would still add another translation and it might be useful to have the generic version if one is being added anyway.

kristastevens commented:
Hello, Krista from Editorial here, reporting for copy review duty! I've got one tiny suggestion to break up the suggested copy into two sentences just to make it a tiny bit easier to parse: 

> This product is an in-app purchase. You can manage it from within the app.

hannahtinkler commented:
> Hello, Krista from Editorial here, reporting for copy review duty! I've got one tiny suggestion to break up the suggested copy into two sentences just to make it a tiny bit easier to parse:
> 
> > This product is an in-app purchase. You can manage it from within the app.

Brilliant, thank you @kristastevens! I've updated it to 'This product is an in-app purchase. You can manage it from within the app store' - switching 'within your app' to 'within the app store' as its editable from the Google Play store or the Apple app store (depending on which was used to purchase it) rather than the WordPress app. Do you think this is clear enough? I'm not sure if the app/app store reference could be confusing.

hannahtinkler committed:
Updates user notice string for clarity.

Relates to #58496

hannahtinkler committed:
Updates user notice string for clarity.

Relates to #58496

hannahtinkler committed:
Disables upgrading from the plans page for IAP purchases.

Relates to #58496

hannahtinkler committed:
Disables upgrading from the plans page for IAP purchases.

Relates to #58496

kristastevens commented:
> Do you think this is clear?

I do! Thanks for asking, @hannahtinkler.

enejb commented:
Nice work Hannah. 

Would it be possible to link out to the app store in question here, since it might be hard to know what app store (Apple or Google) you bought the product from especially it has been a while since you bought the product?

martian77 commented:
Tested as per instructions, and all looked as expected. Code looks ok to me! 👍

hannahtinkler commented:
> Nice work Hannah.
> 
> Would it be possible to link out to the app store in question here, since it might be hard to know what app store (Apple or Google) you bought the product from especially it has been a while since you bought the product?

Thanks @enejb! 😄

For linking, I'm not 100% sure how this would work - we could link to the generic app store in browser but you can't manage anything from there really (I don't even think you can login). As far as I'm aware we can't deep-link to subscriptions in the app store (which is ideally where we'd want a URL to go to I think?), and there are Apple guides for managing subscriptions but I'm not sure if we'd want to link out to those or if there's risk that the link might fail if the URL changes. 

We _could_ send an additional parameter from the API containing the app store type, which would allow us to say something like "You can manage it from within the Apple app store" or You can manage it from within the Google Play app store". I'm not sure if there are other options I've missed...what do you think?

enejb commented:
I think if we can be a bit more specific that would really help our users. So showing Apple App Store vs Google Play store would already be helpful. 

Since deep linking is not really possible. Do the users have to do the payment on their device instead of continuing on the web? 

Can we provide them with a bit more guidance? I don't really know if this is possible or desired, but if we can show them with a QR code or email or SMS them a link, anything like that will help them continue on the path to managing the subscription would be great. ( We do some of that already to get folks to download/install the app - see P7jrAc-i-p2 )

Also, think of these suggestions as possible future enhancements and not something to hold up the PR :)

DavidRothstein commented:
Overall looks good to me.

I agree with providing more guidance beyond "You can manage it from within the app store".  It seems to me like this could use a dedicated section at https://wordpress.com/support/manage-purchases/ (or maybe even an entire doc page of its own) and then the text here could link the user to that page for more detailed help on how to find their purchase in the app store?

One thing stood out to me from the screenshots:

![auto-renewal](https://user-images.githubusercontent.com/235183/143949357-3f66ef18-29a8-433e-84e6-3e138f7caf55.png)

Do we really want to show auto-renew as specifically "on" or "off" when the entire renewal behavior is actually managed via the app store?

The other thing I noticed was that there still seem to be some places in Calypso that let you renew/upgrade these subscriptions.  I am not sure the best way to tackle them all, so I wonder if it will also make sense to add something in the shopping cart itself that catches these cases after the user has tried to add the product to the cart and explains there why you can't do it.

Here's what I found:

You can still upgrade a monthly Jetpack subscription to yearly from the bottom of the manage purchase page:

![jetpack-monthly-purchase-page](https://user-images.githubusercontent.com/235183/143949691-f69621fb-a5db-4c70-b1bd-28b7fae604f9.png)

You can still upgrade to yearly or upgrade to a higher product from the Jetpack plans page (`/plans/[site]`):

![jetpack-plans-page](https://user-images.githubusercontent.com/235183/143949692-6d9d9171-650f-4a27-a15e-6879f374b837.png)

From the WordPress.com plans page, you can't upgrade, although it also leaves you at a bit of a dead end since there's no action or documentation at all regarding how you could upgrade via the app store:

![wpcom-plans-page](https://user-images.githubusercontent.com/235183/143949698-e9de0ee0-33dc-4b94-95b4-d1716f5ec915.png)

On the "My Plan" page (`plans/my-plan/[site]`) there is a "Renew now" link for a subscription expiring soon -- it doesn't actually let you renew (since it just takes you to the manage purchase page rather than checkout) so maybe this is OK as is, although a bit of a dead end given that the manage purchase page isn't going to let you renew it: 
![my-plan-page](https://user-images.githubusercontent.com/235183/143949695-ce679df8-c8c9-40a4-9dab-07f0e0ac34fd.png)

DavidRothstein commented:
>Do we really want to show auto-renew as specifically "on" or "off" when the entire renewal behavior is actually managed via the app store?

Similar question also about how this appears in the list at https://wordpress.com/me/purchases (since one of the columns on that page currently shows the future renewal/expiration behavior of the subscription) -- I am not sure if that should be modified in some way for in-app purchases.

hannahtinkler committed:
Adds ability to accept management link from API and display in puchases area.

Relates to #58496

hannahtinkler committed:
Adds ability to accept management link from API and display in puchases area.

Relates to #58496

hannahtinkler committed:
Hides auto-renew box if the purchase is an in-app purchase.

Relates to #58496

hannahtinkler committed:
Hides auto-renew box if the purchase is an in-app purchase.

Relates to #58496

hannahtinkler committed:
Hide 'upgrade to yearly billing' button for IAP purchases.

Relates to #58496

hannahtinkler committed:
Hide 'upgrade to yearly billing' button for IAP purchases.

Relates to #58496

hannahtinkler committed:
Merge branch 'trunk' into add/in-app-purchase-support

hannahtinkler committed:
Removes 'renew now' button in plan list.

Relates to #58496

hannahtinkler committed:
Removes 'renew now' button in plan list.

Relates to #58496

hannahtinkler committed:
Disables 'get X' button on Jetpack plans if locked plan is already owned.

Relates to #58496

hannahtinkler committed:
Disables 'get X' button on Jetpack plans if locked plan is already owned.

Relates to #58496

hannahtinkler committed:
Merge branch 'trunk' into add/in-app-purchase-support

mattgawarecki commented:
@hannahtinkler Is this PR still active? I'm happy to help -- is there anything you need to continue?

hannahtinkler committed:
Adds IAP directing noticed to plans and my purchases lists.

Relates to #58496

hannahtinkler committed:
Adds IAP directing noticed to plans and my purchases lists.

Relates to #58496

hannahtinkler commented:
> @hannahtinkler Is this PR still active? I'm happy to help -- is there anything you need to continue?

It is 😄  I was just trying to figure out a weird failure on the unit test suite but it might be a known issue, so ready to pass it back to reviewers for now. Thanks!

hannahtinkler commented:
> I think if we can be a bit more specific that would really help our users. So showing Apple App Store vs Google Play store would already be helpful.
I've managed to track down direct links to the stores, so am passing this link to Calypso now and making the notice clickable 😄

@DavidRothstein thanks for pointing out these additional areas! I've updated the PR to:
* Not show the auto-renew/expiry block in Manage Purchases
* Hide the 'Upgrade to yearly billing' button
* Disable the 'Get X' button on the Jetpack plan upgrades, adding 'You can upgrade to Complete through the Jetpack app' as the disabled message
* Added a notice to the plans page to indicate the user can upgrade through the WordPress app
* Replaced the expiry date on me/purchases for IAP purchases with a 'manage it in the app store' notice

I also _think_ I've removed the 'Renew now' button but I don't think I can test this - IAP purchases shouldn't be renewable as they don't have stored payment details against them. For this example, did you amend a regular purchase to have an IAP payment method? This might explain why it was showing, but it _should_ be removed now in any case 😅  I've tested against a manually modified payload but it's not 100% organic.

hannahtinkler committed:
Adds new properties to the defined purchase type.

Relates to #58496

hannahtinkler committed:
Adds new properties to the defined purchase type.

Relates to #58496

DavidRothstein commented:
>IAP purchases shouldn't be renewable as they don't have stored payment details against them. For this example, did you amend a regular purchase to have an IAP payment method?

I may have.... although there's no reason we can't have a "Renew now" option for a subscription with no payment method attached (you get to choose the payment method during checkout anyway).  So it's probably a bug in Calypso that it's not showing "Renew now" more generally there.  Either way, it looks like you fixed it here to prevent it from ever showing up for IAP purchases, regardless of what happens in the future.

All the others look like they're fixed now too - thanks!  I didn't really review the code, but from what I saw skimming through it, I didn't see any problems there either.

jsnajdr commented on client/me/purchases/manage-purchase/index.jsx:
```diff
@@ -979,5 +986,7 @@ export default connect( ( state, props ) => {
 		relatedMonthlyPlanPrice,
 		isJetpackTemporarySite: purchase && isJetpackTemporarySitePurchase( purchase.domain ),
+		purchaseIsInAppPurchase: isInAppPurchase( state, purchase?.id ),
+		purchaseManagementIsLocked: isPurchaseManagementLocked( state, purchase?.id ),
```
I think these two selectors are redundant and don't need to exist at all. You can instead do a direct `purchase?.isLocked` or `purchase?.isInAppPurchase` check.

hannahtinkler replied:
Ah yes, updated!

jsnajdr commented on client/me/purchases/manage-purchase/index.jsx:
```diff
@@ -906,6 +911,8 @@ class ManagePurchase extends Component {
 						isProductOwner={ isProductOwner }
 						getAddNewPaymentMethodUrlFor={ getAddNewPaymentMethodUrlFor }
+						purchaseIsInAppPurchase={ purchaseIsInAppPurchase }
+						purchaseManagementIsLocked={ purchaseManagementIsLocked }
```
You already pass a `purchase` prop to the `PurchaseNotice` component, so the component can figure out the two flags by looking at `purchase`. Doesn't need two new props.

hannahtinkler replied:
Good idea, I'll do that!

jsnajdr commented on client/my-sites/plans/current-plan/purchases-listing.jsx:
```diff
@@ -203,7 +203,17 @@ class PurchasesListing extends Component {
+			( carry, originalPurchase ) =>
+				originalPurchase.id === purchase.id ? originalPurchase.isLocked : carry,
+			false
+		);
```
Can be written without `reduce` as:
```js
this.props.purchases.some( p => p.id === purchase.id && p.isLocked )
```

hannahtinkler replied:
Nice, thank you!

jsnajdr commented on client/my-sites/plans/jetpack-plans/product-card/index.tsx:
```diff
@@ -78,6 +79,16 @@ const ProductCard: React.FC< ProductCardProps > = ( {
+	const existingPurchaseIsIapPurchase = purchases.reduce(
+		( carry, purchase ) => ( purchase.isInAppPurchase ? true : carry ),
+		false
+	);
```
Both conditions can be rewritten into a simpler form with `purchases.some( ... )`.

hannahtinkler replied:
Updated 😅

jsnajdr commented on client/my-sites/plans-features-main/index.jsx:
```diff
@@ -501,6 +515,7 @@ export default connect(
 	( state, props ) => {
 		const siteId = get( props.site, [ 'ID' ] );
 		const currentPlan = getSitePlan( state, siteId );
+		const currentPurchase = getCurrentPlan( state, siteId );
```
Does this really work? `getCurrentPlan` returns the site's _plan_, but we're treating it as _purchase_, which is a quite different object? I can purchase the same plan once as IAP, once with another method. It will be two different purchases, but one plan.

hannahtinkler replied:
Yep - this function uses data that is populated from `/rest/v1.3/sites/{site_id}/plans`. That endpoint grabs data from a bunch of places and (interestingly) lists the subscription ID related to the current plan for the `ID` property 😅  e.g.
```
{
  (plan_id): {
    ID: (subscription_id),
    // other properties
  },
}
```

jsnajdr replied:
I see it now: the `plan.id` field refers to the purchase ID. In that case, however, the `currentPurchase` variable name is incorrect because the object is a plan, not a purchase. You rather want to do something like:
```js
const sitePlan = getSitePlan( state, siteId ); // short version of the plan object
const currentPlan = getCurrentPlan( state, siteId ); // full version of the plan object
const currentPurchase = getByPurchaseId( state, currentPlan?.id ); // purchase of the plan
const isInAppPurchase = currentPurchase?.isInAppPurchase;
```

hannahtinkler replied:
Gotcha, updated! 😄

jsnajdr commented on client/state/selectors/can-upgrade-to-plan.js:
```diff
@@ -28,6 +29,8 @@ export default function ( state, siteId, planKey ) {
 			: PLAN_FREE;
 	const plan = getCurrentPlan( state, siteId );
 
+	const planUpgradeIsLocked = isPurchaseManagementLocked( state, plan?.id );
```
Here again we're treating a `plan` object as a purchase, does the plan really have an `is_locked` field? Also, in my testing `plan.id` is always `null` so the selector will always return `false`?

hannahtinkler replied:
I would expect `plan.id` to be empty either until all the page data had loaded, or if you didn't currently have a subscription to upgrade from (as this ID is strangely the subscription ID, mentioned above), e.g. for a site using the free plan, or for a Jetpack site with no paid Jetpack plans/subscriptions on it. Does this match your use case here or are you seeing something different?

hannahtinkler committed:
Merge branch 'trunk' into add/in-app-purchase-support

hannahtinkler committed:
Replaces some use of `.reduce` and removes some unecessary uses of selectors.

Relates to #58496

hannahtinkler committed:
Replaces some use of `.reduce` and removes some unecessary uses of selectors.

Relates to #58496

jsnajdr commented on client/lib/purchases/types.ts:
```diff
@@ -19,6 +19,8 @@ export interface Purchase {
 	subscriptionStatus: string;
 	domain: string;
+	isLocked: boolean;
+	isInAppPurchase: boolean;
```
Should there be a type for the `iapPurchaseManagementLink` property, too? An optional `string`.

hannahtinkler replied:
Yep, missed that - have added it!

jsnajdr commented on client/me/purchases/manage-purchase/notices.jsx:
```diff
@@ -985,6 +985,25 @@ class PurchaseNotice extends Component {
+						},
+					}
+				) }
+			></Notice>
```
syntax nit: better write a self-closing `<Notice />` JSX.

hannahtinkler replied:
Updated 😄

jsnajdr commented on client/me/purchases/manage-purchase/purchase-meta.jsx:
```diff
@@ -415,7 +415,11 @@ function PurchaseMetaExpiration( {
-	if ( ! purchase || isDomainTransfer( purchase ) ) {
+	const purchaseIsInAppPurchase = useSelector( ( state ) =>
+		isInAppPurchase( state, purchase?.id )
+	);
```
This selector call should also be removed in favor of `purchase?.isInAppPurchase`.

hannahtinkler replied:
Updated 👍

jsnajdr commented on client/my-sites/plans-features-main/index.jsx:
```diff
@@ -513,6 +528,7 @@ export default connect(
 		}
 
 		return {
+			currentPurchaseIsInAppPurchase: isInAppPurchase( state, currentPurchase?.id ),
```
Let's avoid the usage of this selector in favor of `getByPurchaseId`.

hannahtinkler replied:
Updated 👍

jsnajdr commented on client/state/purchases/selectors/index.js:
```diff
@@ -15,6 +15,8 @@ export { getPurchasesError } from './get-purchases-error';
 export { getSitePurchases } from './get-site-purchases';
 export { getUserPurchases } from './get-user-purchases';
+export { isInAppPurchase } from './is-in-app-purchase';
+export { isPurchaseManagementLocked } from './is-purchase-management-locked';
```
It should be easy to completely avoid these selectors by calling `getByPurchaseId( state, id )?.isLocked` etc.

hannahtinkler replied:
You're right - I've removed these selectors and replaced usage with references to the purchase object instead 😄

jsnajdr commented on client/state/purchases/test/selectors.js:
```diff
@@ -114,6 +116,7 @@ describe( 'selectors', () => {
 				refundAmount: NaN,
 				refundOptions: undefined,
 				refundText: undefined,
+				regularPriceText: undefined,
```
Does this PR do anything new about the `regularPriceText` field?

hannahtinkler replied:
No, but the tests were failing without them 🙂

jsnajdr replied:
For me the tests continue to pass even if I remove this line. If tests were failing for you at some point during development, it was probably for some other reason. Let's remove this change as it's not relevant to the PR.

hannahtinkler committed:
Updates selector use with direct access opn purchase object, minor syntax corrections.

Relates to #58496

hannahtinkler committed:
Updates selector use with direct access opn purchase object, minor syntax corrections.

Relates to #58496

jsnajdr reviewed (approved):
Looks good :+1: Thanks for working on this and for addressing all feedback :shipit:

hannahtinkler committed:
Removes unneeded selector property from tests

Relates to #58496

hannahtinkler committed:
Removes unneeded selector property from tests

Relates to #58496

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7160896).

hannahtinkler committed:
Rebases onto trunk and removes a block which has been replaced in trunk.

Relates to #58496

hannahtinkler committed:
Merge branch 'add/in-app-purchase-support' of github.com:Automattic/wp-calypso into add/in-app-purchase-support
